01-ruby-fundamentals/data-structures.md

```diff
@@ -1,211 +1,312 @@
+students.push('Grace')
+students << 'Edith'
+puts "#{students}"
+# => ['Ada', 'Katherine', 'Grace', 'Edith']
+
+puts students[2]
+# => 'Grace'
+
+# Change the element at index 2
+students[2] = 'Anita'
+puts students[2]
+# => 'Anita'
+
+puts "#{students}"
+# => ['Ada', 'Katherine', 'Anita', 'Edith']
+
+# Read an element not in the array
+puts students[10]
+# => nil
+
+puts students[4]
+# => nil
+# Why? Aren't there 4 elements?
 ```
-| Loop #  |   n   |   total   |
-|---------|-------|-----------|
-| 1       |  300  |    300    |
-| 2       |  250  |    550    |
-| 3       |  128  |    678    |
-| 4       |  64   |    742    |
-| 5       |  3016 |    3758   |
 
+#### Iteration
 
-This loop does the same thing with slightly different syntax.  
+- Iteration is the process of doing something to each element in the array, like printing it to the command line or adding 7 to it
+- To iterate manually, you can use a `times` loop on the length of the array
+- There is an `Array#each` method, which is more readable
 
 ```ruby
-deposits = [300, 250, 128, 64, 3016]
+# Print every student to the command line
+students.length.times do |i|
+  puts "#{students[i]}: an excellent student"
+end
 
-total = 0
-# in this loop n becomes 0 to 4 (inclusive) and is used as the "index" of the array.
-(0..4).each do |n|
-	total += deposits[n]
+# Simpler syntax for doing the same thing
+# Note that our iteration variable is the
+# element itself, not the index
+students.each do |student|
+  puts "#{student}: a grand student"
 end
 ```
-| Loop #  |   n   | deposits[n] |   total   |
-|---------|-------|-----------|--------------
-| 1       |  0    |    300    |   300       |
-| 2       |  1    |    250    |   550       |
-| 3       |  2    |    128    |   678       |
-| 4       |  3    |    64     |   742       |
-| 5       |  4    |    3016   |   3758      |
 
+#### Questions
 
-**Question**: How could you write it with the `times` type of loop?
+1. What different ways are there to access the last element in an array?
+1. Given an array with the age of each person in the class,
+    - How would you find the average age?
+    - How would you find the biggest or smallest age?
 
 ### Hashes
 
-A Hash is similar to an array in that it's a collection of data.  However to access an individual element in an Array you always use a number as a key to get to that element.
+Here are some facts about hashes.
+
+#### Creation
+
+- Hashes are created by putting curly braces `{}` around a list of `key: value` pairs
+  - The list can be empty
+- Since a hash is an object, we can store it in a variable
+- You can get the number of key/value pairs with `Hash#length`
+- The hash does **not** keep track of the order of its key/value pairs
 
-So for this list of Pokemon:
 ```ruby
-my_list = [ "Bulbasaur", "Ivysaur", "Venusaur", "Charmander", "Charmeleon", "Charizard", "Squirtle", "Wartortle" ]
-```
-The Array has the following index/value pairs.
+fruit_prices = {
+  apple: 2.15,
+  pear: 3.02,
```

Comment:
Do we prefer the trailing comma? I feel like do in JS but not in Ruby

Reply:
https://github.com/bbatsov/ruby-style-guide#no-trailing-array-commas

Reply:
Good question. This is why we need a style guide.

Reply:
Rubocop says no by default, but you can make it OK for multiline literals
https://github.com/bbatsov/rubocop/issues/713

Reply:
Crap chris had already answered this when I hit send. I guess our conclusions are the same - I'll remove the trailing commas.